x-conduit/src/X/Data/Conduit/Either.hs

```diff
@@ -0,0 +1,168 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module X.Data.Conduit.Either
+  ( ErrorCollectionParams (..)
+  , ErrorStatistics (..)
+  , bindEither
+  , errorCollector
+  , errorConsume
+  , mapRightE
+  , mapRightM
+  , renderErrorStatistics
+  , updateErrorStatistics
+  ) where
+
+import           Control.Monad (Monad)
+import           Control.Monad.IO.Class (MonadIO, liftIO)
+
+import           Data.Conduit (Conduit, Sink)
+import qualified Data.Conduit as DC
+import qualified Data.Conduit.List as DCL
+
+import qualified Data.List as DL
+import           Data.Monoid ((<>))
+
+import           Data.Text (Text)
+import qualified Data.Text as T
+
+
+-- Conduit combinators for working on conduits containing `Either`s.
+
+-- | Standard monadic bind specialized for Either in a conduit.
+bindEither :: Monad m => (a -> Either e b) -> Conduit (Either e a) m (Either e b)
+bindEither = DCL.map . (=<<)
+
+-- | mapRight ef fab : Map a pure function that may produce `Either` over the
+-- `Right` of the input `Either` converting `Left` from the input to `Left` of
+-- the output type.
+-- This is a more general version of `bindEither`.
+mapRightE ::  Monad m => (e -> f) -> (a -> Either f b) -> Conduit (Either e a) m (Either f b)
+mapRightE ef fab = DC.awaitForever $ DC.yield . either (Left . ef) fab
+
+
+-- | Conduit component that allows the `Right` to be say written to a file
+-- while the `Left`s (eg warnings) are passed along the conduit.
+mapRightM :: MonadIO m => (a -> IO (Either e ())) -> Conduit (Either e a) m (Either e a)
+mapRightM actionR =
+  DC.awaitForever $
+    either (DC.yield . Left)
+        (\a -> liftIO (actionR a) >>= either (DC.yield . Left) (const $ DC.yield (Right a)))
+
+
+data ErrorCollectionParams = ErrorCollectionParams
```

Comment:
I wonder whether this belongs here or elsewhere (own library perhaps)? 
The functions above are nice and general, while the ones below are perhaps more specific to how we want to handle the eithers.